clustering/clustering.py

```diff
@@ -0,0 +1,446 @@
+            except ValueError:  # node does not exist in Gaia dataset
+                graph.remove_node(sound_id)
+
+        # Remove isolated nodes
+        graph.remove_nodes_from(list(nx.isolates(graph)))
+
+        return graph
+
+    def create_common_nn_graph(self, sound_ids_list, features=clust_settings.DEFAULT_FEATURES):
+        """Creates a Common Nearest Neighbors Graph representation of the given sounds.
+
+        Args:
+            sound_ids_list (List[str]): list of sound ids.
+            features (str): name of the features to be used for nearest neighbors computation. 
+                Available features are listed in the clustering settings file.
+
+        Returns:
+            (nx.Graph): NetworkX graph representation of sounds.
+        """
+        # first create a knn graph
+        knn_graph = self.create_knn_graph(sound_ids_list, features=features)
+
+        # create the common nn graph
+        graph = nx.Graph()
+        graph.add_nodes_from(knn_graph.nodes)
+
+        for i, node_i in enumerate(knn_graph.nodes):
+            for j, node_j in enumerate(knn_graph.nodes):
+                if j > i:
+                    num_common_neighbors = len(set(knn_graph.neighbors(node_i)).intersection(knn_graph.neighbors(node_j)))
+                    if num_common_neighbors > 0:
+                        graph.add_edge(node_i, node_j, weight=num_common_neighbors)
+
+        # keep only k most weighted edges
+        k = int(np.ceil(np.log2(len(graph.nodes))))
+        for node in graph.nodes:
+            ordered_neighbors = sorted(list(graph[node].iteritems()), key=lambda x: x[1]['weight'], reverse=True)
+            try:
+                neighbors_to_remove = zip(*ordered_neighbors[k:])[0]
+                graph.remove_edges_from(zip([node]*len(neighbors_to_remove), neighbors_to_remove))
+            except IndexError:
+                pass
+
+        # Remove isolated nodes
+        graph.remove_nodes_from(list(nx.isolates(graph)))
+
+        return graph
+
+    def cluster_graph(self, graph):
+        """Applies community detection in the given graph.
+
+        Uses the Louvain method to extract communities from the given graph.
+
+        Args:
+            graph (nx.Graph): NetworkX graph representation of sounds.
+
+        Returns:
+            Tuple(Dict{Int: Int}, int, List[List[Int]], float): 4-element tuple containing the clustering classes for each sound 
+                {<sound_id>: <class_idx>}, the number of communities (clusters), the sound ids in the communities and
+                the modularity of the graph partition.
+        
+        """ 
+        # Community detection in the graph
+        classes = com.best_partition(graph)
+        num_communities = max(classes.values()) + 1
+        communities = [[key for key, value in classes.iteritems() if value==i] for i in range(num_communities)]
```

Comment:
this should be `value == i` - if you can, get pycharm or something to do a style cleanup of these files.

iteritems - use six or .items